Add specification for the Attribute model (#28)

mliszcz opened the pull request:
This PR adds specification for the Attribute model (see #28).

TODO / WIP:
* forwarded attributes

Open points for discussion:
* which RFC shall provide spec for polling? Maybe we should have a dedicated RFC. Polling is configured at device level and can be also applied to commands,
* which RFC shall describe attribute properties (e.g. "min_alarm")? I've seen some work for such properties already done in Property RFC,
* should we have a dedicated RFC for aliases? To define attribute alias I need concepts like "full object name" and "attribute proxy".

mliszcz committed:
Add specification for the Attribute model (#28)

mliszcz committed:
Add aliases to Attribute specification

mliszcz committed:
Add forwarded attribute to Attribute specification

mliszcz committed:
Add event related properties to Attribute specification

hardion commented:
> Open points for discussion:
> 
> * which RFC shall provide spec for polling? Maybe we should have a dedicated RFC. Polling is configured at device level and can be also applied to commands,

Polling mechanism can be described in the RFC/Cache System

> * which RFC shall describe attribute properties (e.g. "min_alarm")? I've seen some work for such properties already done in Property RFC,

IHMO Attribute RFC is the right place since alarm system concern 99% the  Tango attribute (some side effect on the Device State which can be describe in the Device RFC)

> * should we have a dedicated RFC for aliases? To define attribute alias I need concepts like "full object name" and "attribute proxy".

Just make a reference to the Device RFC.

mliszcz committed:
Reference Device RFC from attribute alias section

mliszcz committed:
Add reference to cache system from attribute polling

mliszcz commented:
Thanks @hardion!
> Polling mechanism can be described in the RFC/Cache System

I have added a note regarding polling and referenced RFC/Cache system.

> IHMO Attribute RFC is the right place since alarm system concern 99% the Tango attribute (some side effect on the Device State which can be describe in the Device RFC)

Already documented.

> Just make a reference to the Device RFC.

Done.

srgblnch commented:
Cześć @mliszcz !

Sorry for the delay in my task.I'm back on line. I'll use letter to tag comments I do about the text so then we can reference them in an easier way.

a) In the goals section, there are some more attribute types* than 'read only' and 'read write', I've never used but there is the possibility to have 'write only' and 'read with write'. You later on mention this but don't fit with the first lines of the rfc.
\* Not the same type than datatype.

b) like in the commands rfc, I suggest to reference the Object Oriented paradigm to describe what an attribute mean in Tango. Just the meaning of the name of Tango does a direct reference to the Object paradigm. I think this reference help then in the User Cases section because it already fits in a common view.

c) In the attribute definition, the name may have more restrictions (like don't contain slashes I think).

d) Continuing with the attribute definition for the name, I think it would be nice if, apart to describe what we already have in the implementation, guide to encourage or discourage unpleasant behaviors like for me it is the empty name.

I'll do another read soon in order to think more in a refining of the rfc.

Thanks for the effort!

./Sergi

mliszcz commented:
Thanks @srgblnch!

> a) In the goals section, there are some more attribute types* than 'read only' and 'read write',

My intention was to not go into details in the goals section. Shall I reword the goals and include all attribute access types?

> b) like in the commands rfc, I suggest to reference the Object Oriented paradigm

Sure, I'll add this information.

> c) In the attribute definition, the name may have more restrictions (like don't contain slashes I think).

I've checked with c++ server and python client. Both `\` and `/` are fine. Maybe some combination of letters is not allowed - I need to have a look at the source code.

```cpp
rw_scalar_doubleAttrib():Attr("~`\1!@#$%^&*()_+{}[]:;'|\\<,>.?/ǢǣǮȌ-=", Tango::DEV_DOUBLE, Tango::READ_WRITE) {};
```

```python
In [6]: filter(lambda a: '!@#' in a.name, dp.attribute_list_query())[0].name
Out[6]: "~`\x01!@#$%^&*()_+{}[]:;'|\\<,>.?/\xc7\xa2\xc7\xa3\xc7\xae\xc8\x8c-="

In [8]: dp.read_attribute("~`\x01!@#$%^&*()_+{}[]:;'|\\<,>.?/\xc7\xa2\xc7\xa3\xc7\xae\xc8\x8c-=").value
Out[8]: 4.6464171418034e-310
```

> d) Continuing with the attribute definition for the name, I think it would be nice if, apart to describe what we already have in the implementation, guide to encourage or discourage unpleasant behaviors like for me it is the empty name.

I've just checked the command and property RFCs. Seems that they also had some problems with naming and ended up with a simple:

```abnf
command-name = 1*VCHAR
```

Maybe we shall do the same for the Attribute? I'd like to avoid giving guidelines in a specification, because the guidelines are subjective.

bourtemb commented:
> > c) In the attribute definition, the name may have more restrictions (like don't contain slashes I think).
> 
> I've checked with c++ server and python client. Both `\` and `/` are fine. Maybe some combination of letters is not allowed - I need to have a look at the source code.
> 
> ```c++
> rw_scalar_doubleAttrib():Attr("~`\1!@#$%^&*()_+{}[]:;'|\\<,>.?/ǢǣǮȌ-=", Tango::DEV_DOUBLE, Tango::READ_WRITE) {};
> ```
> 
> ```python
> In [6]: filter(lambda a: '!@#' in a.name, dp.attribute_list_query())[0].name
> Out[6]: "~`\x01!@#$%^&*()_+{}[]:;'|\\<,>.?/\xc7\xa2\xc7\xa3\xc7\xae\xc8\x8c-="
> 
> In [8]: dp.read_attribute("~`\x01!@#$%^&*()_+{}[]:;'|\\<,>.?/\xc7\xa2\xc7\xa3\xc7\xae\xc8\x8c-=").value
> Out[8]: 4.6464171418034e-310
> ```

It is true that the library lets you create some attributes with these strange characters but POGO doesn't. Please refer to :
https://github.com/tango-controls/pogo/blob/0887143c96f95576a242f69d7f953db72ba93eea/org.tango.pogo.pogo_gui/src/org/tango/pogo/pogo_gui/tools/Utils.java#L404-L436

In the original Tango philosophy, Device servers programmers had to use POGO to define their attributes. Then the dynamic attribute feature got implemented and it seems like no checks were added to the library. I think it is a mistake/inconsistency.

> 
> > d) Continuing with the attribute definition for the name, I think it would be nice if, apart to describe what we already have in the implementation, guide to encourage or discourage unpleasant behaviors like for me it is the empty name.
> 
> I've just checked the command and property RFCs. Seems that they also had some problems with naming and ended up with a simple:
> 
> ```abnf
> command-name = 1*VCHAR
> ```

Please note that what is proposed in the property RFC is not definitive. It is a draft. We simply didn't have enough time to come up with a more precise abnf definition yet.

> 
> Maybe we shall do the same for the Attribute? I'd like to avoid giving guidelines in a specification, because the guidelines are subjective.

mliszcz committed:
Refer to object-oriented design in Attribute spec

mliszcz committed:
Add recommendations for Attribute name

mliszcz commented:
Hi @srgblnch 
I think I resolved:
b) - just one extra sentence referring to the OOP in the *Goals* section
and c) + d) - added a note with RECOMMENDations for attribute naming - @bourtemb can you also comment if it's ok for you now? I think use of Pogo should not be mandatory (e.g. in Python HL there is almost no boilerplate code to generate).

bourtemb commented on 4/README.md:
```diff
@@ -0,0 +1,346 @@
+  It MUST be unique [^1] among all Attributes of a particular device.
+  It MAY contain any character (including multi-byte characters)
+  except that it MUST NOT contain `\0` character.
+  It MUST not not be "Status" or "State", which are reserved names.
```
State and Status are actually valid attribute names so I think this sentence is not correct... It is just that the device server programmer can not create additional attributes having these names since this would contradict the first point above: "It MUST be unique among all Attributes of a particular device".

mliszcz replied:
Proposed fix: remove "It MUST not not be "Status" or "State", which are reserved names."

bourtemb commented on 4/README.md:
```diff
@@ -0,0 +1,346 @@
+## Specification
+
+An Attribute has:
+* A set of static metadata that constitute to Attribute's definition,
```
This sentence does not sound very English to me.
A set of static metadata that constitute Attribute's definition ?
A set of static metadata that contribute to Attribute's definition ?
...?

mliszcz replied:
Proposed fix: "A set of static metadata that constitute Attribute's definition".

By "attribute definition" I meant all the parameters (name, type, etc...) you configure statically in Pogo (or pass in `Tango::Attr` constructor) and cannot be changed later at runtime.

bourtemb commented on 4/README.md:
```diff
@@ -0,0 +1,346 @@
+  It MAY contain any character (including multi-byte characters)
+  except that it MUST NOT contain `\0` character.
+  It MUST not not be "Status" or "State", which are reserved names.
+  It MAY be empty.
```
It MAY be empty but SHOULD NOT be empty.
I am not convinced we should even put a sentence like "It MAY be empty" when we know that it won't work correctly in this use case. I would remove this sentence to avoid confusion.

mliszcz replied:
Below is a note that there should be at least one character in attribute's name. Proposed fix: I'll remove "It MAY be empty."

bourtemb commented on 4/README.md:
```diff
@@ -0,0 +1,346 @@
+* *name*, a string identifying the Attribute.
+  It MUST be unique [^1] among all Attributes of a particular device.
+  It MAY contain any character (including multi-byte characters)
+  except that it MUST NOT contain `\0` character.
```
I find this '\0' detail quite confusing for the C Binding client implementation for instance...

mliszcz replied:
You can't have \0 (since it's end of string), but you can have  other bytes like \1, \2, ... But - such names will not be accepted by Pogo (which expects only ASCII letters and numbers).

I can remove the note about \0 ("except that it MUST NOT contain `\0` character.")

Do we want to remove the "any character" part as well? - "It MAY contain any character (including multi-byte characters)". Even though it will not work with Pogo, I think non-English letters in attribute names are still a valid use case.

bourtemb replied:
> You can't have \0 (since it's end of string), but you can have other bytes like \1, \2, ... But - such names will not be accepted by Pogo (which expects only ASCII letters and numbers).
> 
> I can remove the note about \0 ("except that it MUST NOT contain `\0` character.")

I don't know what would be the best. When reading it I just found it confusing:
> *name*, a **string** identifying the Attribute.
> it MUST NOT contain `\0` character.

A C string contains a `\0` at the end by definition so this seemed a bit contradictory to me.
 
> Do we want to remove the "any character" part as well? - "It MAY contain any character (including multi-byte characters)". Even though it will not work with Pogo, I think non-English letters in attribute names are still a valid use case.

Non-English letters/multi-byte characters in attribute names can be a valid use case indeed.  There might be some precautions needed for this to work correctly though. I think character set encoding configuration is important in this case on the database side , on the CORBA side and/or on the OS side.
I don't know whether some users are already doing it. This use case is not tested in cppTango I think.

mliszcz replied:
So I've removed the confusing note about the `\0`. I've left the information about multi-byte characters.
* Tango should work (with minor tweaking, see [TangoDatabase#16](https://github.com/tango-controls/TangoDatabase/issues/16)) with any Database chacter set.
* According to the answers in [this stackoverflow question](https://stackoverflow.com/questions/402283/stdwstring-vs-stdstring), `std::string` can hold unicode.
* Simple test shows it is possible to read attribute with non-ASCII name.

bourtemb commented on 4/README.md:
```diff
@@ -0,0 +1,346 @@
+An Attribute MUST have associated following static metadata:
+* *memorized*, a flag describing whether *set value* is stored in the database
+  and restored during Attribute initialization,
+* *write hardware at init*, a flag describing whether *set value* is
```
I would use "memorized set value" instead of "set value" here.

mliszcz replied:
I'll do as you propose

bourtemb commented on 4/README.md:
```diff
@@ -0,0 +1,346 @@
+
+To set *memorized*, following constraints MUST be satisfied:
+* *data format* MUST be *SCALAR*,
+* *data type* MUST NOT be *STATE* or *ENCODED*,
```
Indeed Attr:set_memorized() method seems to forbid memorized attributes of type STATE:
https://github.com/tango-controls/cppTango/blob/f3efe02db1ef1bfd5347901f0e7b5e20aba478da/cppapi/server/attrdesc.cpp#L665-L675

I think this is a mistake in the current implementation. Attributes of type DevState (other than attribute named "State") were not there at the beginning when attributes were created in Tango. It looks like the test in Attr::set_memorized() has not been updated when the support of DevState user attributes was added.
In theory, there is no reason not to support STATE memorized attributes and we can even see that some other methods are ready for that, like Device_3Impl::write_attributes_in_db for instance:
https://github.com/tango-controls/cppTango/blob/tango-9-lts/cppapi/server/device_3.cpp#L2296-L2382

So, thank you @mliszcz for this discovery. We should fix that in cppTango (and add a test to test State memorized attributes).

mliszcz replied:
So, do you propose to remove this constrains from the RFC? Or shall I add a note here that this will be changed in the future?

bourtemb replied:
> 
> 
> So, do you propose to remove this constrains from the RFC? Or shall I add a note here that this will be changed in the future?

This constraint on the type (STATE or ENCODED) might not exist in the latest version of JTango.

JTango does not (no longer) support(s) spectrum or image memorized attributes mainly due to performance issues:

https://github.com/tango-controls/JTango/blob/jtango-9-lts/server/src/main/java/org/tango/server/attribute/AttributeImpl.java#L482-L492

If an implementation manages to support these kinds of memorized attributes in an efficient manner, I would not forbid it.

We could add a note about what kind of memorized attributes are currently supported by JTango and cppTango.

In any case, what is important here is the concept of memorized attribute.

bourtemb commented on 4/README.md:
```diff
@@ -0,0 +1,346 @@
+* *writable* MUST NOT be *READ* or *READ_WITH_WRITE*.
+
+If *memorized* is set and if database is being used,
+the *set value* MUST be persisted in the *__value* property upon each write
```
*__value* property  => *__value* attribute property

mliszcz replied:
Will change as proposed.

bourtemb commented on 4/README.md:
```diff
@@ -0,0 +1,346 @@
+the *set value* MUST be persisted in the *__value* property upon each write
+to the Attribute.
+
+If *memorized* is set and *__value* property is other than "Not used yet"
```
*__value* property  => *__value* attribute property

mliszcz replied:
Will change.

bourtemb commented on 4/README.md:
```diff
@@ -0,0 +1,346 @@
+If *memorized* is set and *__value* property is other than "Not used yet"
+(default):
+* only during Attribute initialization (at startup):
+  * *set value* MUST be set with a value stored in *__value* property,
```
*__value* property  => *__value* attribute property

bourtemb commented on 4/README.md:
```diff
@@ -0,0 +1,346 @@
+#### Forwarded attribute
+
+An Attribute can be a forwarded attribute. A forwarded Attribute MUST have
+associated *__root_att* property.
```
*__root_att* property => *__root_att* attribute property

bourtemb commented on 4/README.md:
```diff
@@ -0,0 +1,346 @@
+  [`printf`](http://man7.org/linux/man-pages/man3/printf.3.html)
+  format string specification,
+* *min value*, describing the minimum value of Attribute's *read value*.
+  It MUST be a vaild numerical value.
```
vaild  => valid
(there are several other occurences of this typo below)

bourtemb commented on 4/README.md:
```diff
@@ -0,0 +1,346 @@
+  format string specification,
+* *min value*, describing the minimum value of Attribute's *read value*.
+  It MUST be a vaild numerical value.
+  It MUST be defined only for Attributes with numerical *data type*.
```
MUST => MAY ?
It can be unspecified

mliszcz replied:
The header says that the Attribute MAY have properties listed below.
By "it MUST ..." I meant "if defined, it MUST ...".

Should I explicit add "if defined" to each requirement?

bourtemb replied:
> Should I explicit add "if defined" to each requirement?

That would be one way to make it less confusing indeed, I think! 
Or maybe having something like:

If defined:
- It MUST ...
- It MUST ...

I am just sharing ideas (I am not the editor ;-) )

bourtemb commented on 4/README.md:
```diff
@@ -0,0 +1,346 @@
+  It MUST be a vaild numerical value.
+  It MUST be defined only for Attributes with numerical *data type*.
+  It MUST be defined only for writable Attributes.
+  It MUST be lower than *max value*,
```
if *max value* is specified

bourtemb commented on 4/README.md:
```diff
@@ -0,0 +1,346 @@
+  representation. It MUST comply to
+  [`printf`](http://man7.org/linux/man-pages/man3/printf.3.html)
+  format string specification,
+* *min value*, describing the minimum value of Attribute's *read value*.
```
I would rather put:
*read value* => *set value*

bourtemb commented on 4/README.md:
```diff
@@ -0,0 +1,346 @@
+  It MUST be defined only for Attributes with numerical *data type*.
+  It MUST be defined only for writable Attributes.
+  It MUST be lower than *max value*,
+* *max value*, describing the maximum value of Attribute's *read value*.
```
Same here, I think *max value* is especially important for the *set/write value*

bourtemb commented on 4/README.md:
```diff
@@ -0,0 +1,346 @@
+  It MUST be a vaild numerical value.
+  It MUST be defined only for Attributes with numerical *data type*.
+  It MUST be defined only for writable Attributes.
+  It MUST be greater than *max value*.
```
It MUST be greater than *max value*. => It MUST be greater than *min value*.

bourtemb commented on 4/README.md:
```diff
@@ -0,0 +1,346 @@
+  It MUST be defined only for Attributes with numerical *data type*.
+  It MUST be greater than *min alarm*,
+* *min warning*, is deprecated,
+* *max warning*, is deprecated,
```
Where did you find the information that *min warning* and *max warning* are deprecated?
The documentation does not mention that: https://tango-controls.readthedocs.io/en/latest/development/device-api/attribute-alarms.html?highlight=min_warning#the-level-alarms

bourtemb commented on 4/README.md:
```diff
@@ -0,0 +1,346 @@
+  *read value* and *set value* after *delta t* time period,
+  above which the attribute is considered as having alarm *quality*.
+  It MUST be defined only for writable Attributes,
+* *delta t*, describing the time period after which the difference between
```
Can we specify the unit of this time period?

mliszcz replied:
Proposed fix: "... describing the time period (**in milliseconds**) after which the difference between ..."

bourtemb commented on 4/README.md:
```diff
@@ -0,0 +1,346 @@
+  or a pair of valid numerical values separated by a `,`.
+  If one value is specified, it MUST be used for both negative
+  and positive change.
+  If two values are specified, the first MUST be used for netative change
```
netative  => negative

bourtemb commented on 4/README.md:
```diff
@@ -0,0 +1,346 @@
+  It MUST be either a valid numerical value
+  or a pair of valid numerical values separated by a `,`.
+  If one value is specified, it MUST be used for both negative
+  and positive change.
```
It would be good to give more specifications about the relative change. How is it expressed?
What does a relative change of 1 means?

mliszcz replied:
Proposed fix:
It MUST be expressed as a precentage change relative to the value reported in the previous event, e.g. a relative change of 1 will generate an event when:
```
(current value - previous value) / previous value > 1%
```

bourtemb commented on 4/README.md:
```diff
@@ -0,0 +1,346 @@
+  in milliseconds after which a *PERIODIC* or *ARCHIVE* event MUST be reported,
+  regardless of Attribute's *read value*. If *period* is not specified,
+  a default value of 1000 ms is used,
+  It MUST be defined only for Attributes with numerical *data type*.
```
>  It MUST be defined only for Attributes with numerical *data type*.

I don't think this is true for the *period* and *archive period*

mliszcz replied:
Correct! Proposed fix: remove "It MUST be defined only for Attributes with numerical *data type*."

bourtemb commented on 4/README.md:
````diff
@@ -0,0 +1,346 @@
+
+```abnf
+rel-change = change
+abs-change = chagne
````
chagne => change (same a bit below)

bourtemb commented on 4/README.md:
```diff
@@ -0,0 +1,346 @@
+number = [ "-" ] 1*DIGIT [ "." ] *DIGIT
+change = number [ "," number ]
+
+period = [ "-" ] 1*DIGIT
```
negative period?

mliszcz replied:
Correct. It's either positive or 0 (default / not set). Proposed fix: `period = 1*DIGIT`

bourtemb commented on 4/README.md:
```diff
@@ -0,0 +1,346 @@
+
+An Attribute MAY have associated *alias*, a string which can be used
+in place of *full attribute name* to address the Attribute.
+An Attribute MUST NOT have more than one *alias*.
```
I think this is a limitation of the current system with jive but I don't see why we could not have several aliases for the same attribute. How did you come to this conclusion? Did you find something in the doc or in the source code or was it simply by trying to define an attribute alias with jive?

mliszcz replied:
This is how it is implemented in the database:
https://github.com/tango-controls/TangoDatabase/blob/master/DataBase.cpp#L6150-L6155

Upon alias definition, you clear all existing aliases. Also see:

```
In [2]: db = tango.Database()

In [3]: db.put_attribute_alias('sys/tg_test/1/long_scalar', 'first_alias_for_test_long_scalar')

In [4]: db.get_attribute_alias_list('*')
Out[4]: DbDatum(name = '*', value_string = ['first_alias_for_test_long_scalar', 'rw_scalar_double-x', '', 'debian8_attr_alias'])

In [5]: db.put_attribute_alias('sys/tg_test/1/long_scalar', 'second_alias_for_test_long_scalar')

In [6]: db.get_attribute_alias_list('*')
Out[6]: DbDatum(name = '*', value_string = ['second_alias_for_test_long_scalar', 'rw_scalar_double-x', '', 'debian8_attr_alias'])
```

bourtemb replied:
You're right indeed! What you wrote is correct! I just don't understand why we have this limitation for the moment. I thought it might be useful in some situations to have several aliases pointing to the same attribute.
Maybe this is not needed.

bourtemb commented on 4/README.md:
```diff
@@ -0,0 +1,346 @@
+
+An *alias* MUST be stored in the database.
+
+An *alias* MUST be unique.
```
in the sense that a given alias name can only be associated to a unique attribute?

mliszcz replied:
It must be unique across the whole database.
Please see DataBase::db_put_attribute_alias.
* First, we reject the new alias if this alias is already in use for another attribute
* Then we delete all aliases for the current attribute:
* And we finally insert the new alias

Effectively, the `alias` column is unique.

bourtemb commented on 4/README.md:
```diff
@@ -0,0 +1,346 @@
+
+In order to send *PERIODIC*, *CHANGE*, *ARCHIVE*, *PIPE* and *DATA READY*
+events, an attribute must be polled. See
+[Attribute polling section](#attribute-events) for more details.
```
This is not true. Events can be pushed by code. The polling is not mandatory in this case.

mliszcz replied:
Proposed fix: replace above text with:

Events can be send automatically by a polling mechanism, automaticallyby the Tango system or manually from the device server code.

The polling mechanism can send *PERIODIC*, *CHANGE* and *ARCHIVE* events.
In order to send events via the polling mechanism, the polling for the Attribute MUST be configured in the Device with *polled attr* Property.
See "Attribute properties" section for conditions upon which the events are sent.
See "RFC XX/Cache system" for more details on polling configuration.

The Tango system MUST send *ATTR CONF* event if Attribute configuration is changed.
See "Attribute properties" section for a list of modifiable attribute Properties.

The device server code can manually send *CHANGE*, *ARCHIVE*, *DATA READY* and *USER* events without the need to configure polling for the Attribute.

bourtemb replied:
> Proposed fix: replace above text with:
> 
> Events can be send automatically by a polling mechanism, automaticallyby the Tango system or manually from the device server code.

can be send => can be sent
automaticallyby  => automatically by 

> 
> The polling mechanism can send _PERIODIC_, _CHANGE_ and _ARCHIVE_ events.
> In order to send events via the polling mechanism, the polling for the Attribute MUST be configured in the Device with _polled attr_ Property.

I would just write something like:
"In order to send events via the polling mechanism, the polling for the Attribute MUST be enabled"
without going into more details. The polling RFC can give all the details on how to configure/start the polling for a given attribute.

> The device server code can manually send _CHANGE_, _ARCHIVE_, _DATA READY_ and _USER_ events without the need to configure polling for the Attribute.

I think we should remove the mention to PIPE events a bit above too. PIPE are not attributes so we should not mention that in this RFC which is dedicated to attributes.

bourtemb commented on 4/README.md:
````diff
@@ -0,0 +1,346 @@
+
+Formal specification of Attribute name is given below:
+```ABNF
+attribute-name = *attribute-name-char
````
I think we should impose at least one character here because it does not work properly in the current system with empty attribute names.

mliszcz replied:
Do you mean that it will be rejected by Pogo or will not work with tools like Jive?
I already added a note:
"Also note that the Attribute name SHOULD contain at least one letter and SHOULD NOT start with a digit."

In my opinion the proper place to put any restrictions on attribute name would be constructor `Attr::Attr`. Otherwise we will have a specification which is not followed by any implementation.

bourtemb replied:
> Do you mean that it will be rejected by Pogo or will not work with tools like Jive?

I mean there will be some unexpected side effects for sure. I didn't try it so I don't know how bad it will behave actually.

> I already added a note:
> "Also note that the Attribute name SHOULD contain at least one letter and SHOULD NOT start with a digit."

Indeed, I saw this. The abnf definition is much more relaxed. Maybe we should add a recommended attribute name abnf definition? I understand your point of view. You're simply documenting the current implementation.
I think we should not tolerate empty attribute names in next major Tango versions.

> 
> In my opinion the proper place to put any restrictions on attribute name would be constructor `Attr::Attr`. 

I agree with you

bourtemb reviewed (commented):
Thanks @mliszcz for this very interesting work. This actually helps to find some issues with the current code.
Please have a look at my comments.

mliszcz committed:
Add Sergi's mail to the header

mliszcz committed:
Refine and clarify some requirements in Attribute spec

mliszcz commented:
Thanks a lot @bourtemb for an in-depth review and a lot of useful suggestions!

@srgblnch, in the lates commit (87f0d0d) I did a rework according to @bourtemb's suggestions. Please check if you are ok with proposed changes.

These still need to be addressed:
* [x] read / write and other access types in "Goals" section
* [x] unicode names
* [x] *memorized* restrictions and differences between Java and C++
* [x] warning deprecation - I do not remember when I have found this, maybe I'm wrong, need to check
* [x] empty names

srgblnch commented:
> > a) In the goals section, there are some more attribute types* than 'read only' and 'read write',
> 
> My intention was to not go into details in the goals section. Shall I reword the goals and include all attribute access types?
> > b) like in the commands rfc, I suggest to reference the Object Oriented paradigm
> 
> Sure, I'll add this information.

I change my mind and I propose that would be better if here, in an introduction level we mention that the concept of a tango attribute comes from the concept of an attribute in object oriented paradigm. This is merging items 'a' and 'b'. Not going much to the detail, as a high light. Later in the 'Attr definition' section is where the current access can be and are described.

about 'c'. Good point the comment about pogo restriction. We may think that pogo is part of the standard so this specification may include it and one day we can move this restriction down (instead of being in the top of our tools).

about 'd'. Yes, it is true that guidelines are subjective and specifications tries to be objective. But when the specification of something that is already implemented supports something that can cause problems, imho is when we can introduce this amendments.

Even if the implementation allows an empty string as attribute name, how we later can reference it because its name collides with the device name. What do you think?

srgblnch commented:
Hi, @bourtemb 
I've seen you reference an open issue in cppTango about this limitation on memorize writable DevState attributes. What I think:
1) as far as we are describing what we have now a days, I think the specification should say that those attributes cannot be memorized. If java implementation allows it, nice, but don't make the cppTango go below the specs.
2) A clear code in cppTango that avoid explicitly DevStates and DevEncodeds, we may have to think during a while why someone did it. Best case, if how wrote it thought they are not necessary, but perhaps there is a hard reason behind.
Imho, We should describe what we have nowadays and avoid to make our implementations not compilant with the specification. So I prefer to specify the limit and extend it when all our implementations extend it.
What do you think?

mliszcz committed:
Remove notes about multi-byte characters

bourtemb commented:
> Hi, @bourtemb
> I've seen you reference an open issue in cppTango about this limitation on memorize writable DevState attributes. What I think:
> 
>     1. as far as we are describing what we have now a days, I think the specification should say that those attributes cannot be memorized.

If you write something like that it means that Java implementation doesn't comply to the specs because it looks like the specs says that it is forbidden or impossible to memorize these kinds of attributes.
Maybe we just need to find the right way to express it. 

>     2. A clear code in cppTango that avoid explicitly DevStates and DevEncodeds, we may have to think during a while why someone did it.

DevEncoded are currently forbidden because this can imply a lot of data because DevEncoded are very often used to store images. So it is currently forbidden for performance reasons.
For the DevState, I guess it is historical but nobody knows why this check is here.

>  Best case, if how wrote it thought they are not necessary, but perhaps there is a hard reason behind.

@taurel does not remember why this check is here for the DevState attributes and doesn't see a good reason to forbid DevState memorized attributes.

>        Imho, We should describe what we have nowadays and avoid to make our implementations not compilant with the specification.

I agree, but we have to take into account the differences between the current implementations (Java, C++,...). I think it is a good exercise to identify these differences too.
The different implementations do not support the features at the same level.
What about the Labview implementation? I don't think we should base our specs on the implementation having the lowest number of features.
I think the main difficulty of this rfc writing exercise is to define precisely what we want to describe.
If we want to describe the current implementation**s**, then, I think maybe we should have different sections for the different implementations because they are different. 
The main concepts are the same, and I think this is what is the most important to be described, but the behaviour can be slightly different between implementations and some features might not be supported in some implementations but in this case we should not write in the specs that these features cannot be implemented or are forbidden.
I'm not against notes specifying the limits of the current implementations though.

> So I prefer to specify the limit and extend it when all our implementations extend it.
>        What do you think?

For me, what is the most important here is the concept of memorized attribute.
I think this feature can be optional in some implementations if performances would be too bad.

mliszcz commented:
Hi @bourtemb and @srgblnch 

How about rewording:
```
To set memorized, following constraints MUST be satisfied:
    data format MUST be SCALAR,
    data type MUST NOT be STATE or ENCODED,
    writable MUST NOT be READ or READ_WITH_WRITE.
```
Into:
```
A conforming implementation MUST allow to set "memorized" flag and support memorization
for attributes which:
    data format is SCALAR,
    data type is not STATE or ENCODED,
    "writable" flag is not READ or READ_WITH_WRITE.
```
This is true both for Java and C++ implementation. If JTango supports other data formats or if cppTango will decide to lift the constraints for DevState types, they will still be in-line with the spec.

The problem is that we should document the features available in various implementations somewhere. Maybe in the documentation? IMO, RFC is not a good place to mention any particular implementation.

bourtemb commented:
> Hi @bourtemb and @srgblnch
> 
> How about rewording:
> [...]
> 
> Into:
> 
> ```
> A conforming implementation MUST allow to set "memorized" flag and support memorization
> for attributes which:
>     data format is SCALAR,
>     data type is not STATE or ENCODED,
>     "writable" flag is not READ or READ_WITH_WRITE.
> ```
> 
> This is true both for Java and C++ implementation. If JTango supports other data formats or if cppTango will decide to lift the constraints for DevState types, they will still be in-line with the spec.
> 

Sounds good to me. I would just replace the last sentence with:
"writable" flag is WRITE or READ_WRITE.

> The problem is that we should document the features available in various implementations somewhere. Maybe in the documentation? 

Indeed the doc could be a good place.

>IMO, RFC is not a good place to mention any particular implementation.

In theory it should not indeed...

mliszcz committed:
Clarify meaning of 'writable' attribute metadata

mliszcz committed:
Define only minimal requirements for memorized Attributes

mliszcz committed:
Require at least one character in Attribute name

mliszcz committed:
Specify meaning of min_ and max_warning Attribute properties

mliszcz commented:
Hi @srgblnch and @bourtemb 

I have addressed the remaining points:

* read / write and other access types in "Goals" section
  * no change needed as discussed with Sergi on slack
* unicode names
  * removed, officially we support only ASCII letters and numbers (and this is what is tested)
* memorized restrictions and differences between Java and C++
  * reworded so that implementations can deliver more features than the specification requires.
* warning deprecation
  * I couldn't find the source. Probably this is my mistake. I added a description for min_ and max_warning.

Please take a look if you are ok with recent changes.
Thanks!
